crates/sui-rosetta/src/types/internal_operation/stake.rs

```diff
@@ -0,0 +1,205 @@
+        budget: Option<u64>,
+    ) -> Result<TransactionObjectData, Error> {
+        let Self {
+            sender,
+            validator,
+            amount,
+        } = self;
+
+        if amount.is_none() {
+            let all_coins = client
+                .coin_read_api()
+                .get_coins_stream(sender, None)
+                .collect::<Vec<_>>()
+                .await;
+
+            let total_sui_balance = all_coins.iter().map(|c| c.balance).sum::<u64>() as i128;
+            let mut iter = all_coins.into_iter().map(|c| c.object_ref());
+            let gas_coins: Vec<_> = iter.by_ref().take(MAX_GAS_COINS).collect();
+            let extra_gas_coins: Vec<_> = iter.collect();
+
+            let budget = match budget {
+                Some(budget) => budget,
+                None => {
+                    let pt = stake_pt(validator, total_sui_balance as u64, true, &extra_gas_coins)?;
+                    budget_from_dry_run(client, pt, sender, gas_price).await?
+                }
+            };
+
+            return Ok(TransactionObjectData {
+                gas_coins,
+                extra_gas_coins,
+                objects: vec![],
+                total_sui_balance,
+                budget,
+            });
+        }
+
+        let amount = amount.expect("We already handled amount: None");
+
+        // amount and budget is given
+        if let Some(budget) = budget {
+            let all_coins = client
+                .coin_read_api()
+                .select_coins(sender, None, (amount + budget) as u128, vec![])
+                .await?;
+            let total_sui_balance = all_coins.iter().map(|c| c.balance).sum::<u64>() as i128;
+
+            let mut iter = all_coins.into_iter().map(|c| c.object_ref());
+            let gas_coins: Vec<_> = iter.by_ref().take(MAX_GAS_COINS).collect();
+            let extra_gas_coins: Vec<_> = iter.collect();
+
+            return Ok(TransactionObjectData {
+                gas_coins,
+                extra_gas_coins,
+                objects: vec![],
+                total_sui_balance,
+                budget,
+            });
+        }
+
+        // amount is given, budget is not
+        let mut coins_stream = Box::pin(client.coin_read_api().get_coins_stream(sender, None));
+        // Fetch it once instead of fetching it again and again in the below loop.
+        let gas_price = match gas_price {
+            Some(p) => p,
+            None => client.governance_api().get_reference_gas_price().await? + 100, // make sure it works over epoch changes
+        };
+
+        let mut all_coins = vec![];
+        let mut gas_coins: Vec<_>;
+        let mut extra_gas_coins: Vec<_>;
+        let mut gathered = 0;
+        let mut budget = START_GAS_UNITS * gas_price;
+        // We need to dry-run in a loop, because depending on the amount of coins used the tx might
+        // differ slightly: (merge / no merge / number of merge-coins)
+        loop {
```

Comment:
the code to determine correct gas for tx with added merge command is quite complex but very similar for pay sui and stake, is it possible to reduce duplicate code?

Reply:
[82c6950](https://github.com/MystenLabs/sui/pull/20283/commits/82c695002cf31c5164d70c838717fc4c7f9d27ef)